db/db_impl_compaction_flush.cc

```diff
@@ -2317,6 +2389,10 @@ Status DBImpl::BackgroundCompaction(bool* made_progress,
         }
       }
     }
+    if (!c) {
+      // There is no valid compaction for picked column family.
+      DecreaseCompactionTasks(cfd, log_buffer);
```

Comment:
I'm a little bit worried that getting the compaction tasks decreased in different places makes the code hard to maintain in the future. It's easy to make mistakes in the future when code is moved around. For example, it is not easy to reason that whether we released something we have required, whether we forget to release the token, or whether we have double released. Can we make the release explicit? For example, when asking for a token, explicitly hold something here, rather than use various condition of `!c` or not. For example, we have an explicit unique_ptr of tokens here: https://github.com/facebook/rocksdb/blob/189f0c27aaecdf17ae7fc1f826a423a28b77984f/db/write_controller.h#L123-L128 so that we only need to remember to release the token before the end of the function. At least we can have a boolean to indicate that whether we have asked for token, and do an assert before where we release it.

Reply:
Thanks for the  suggestion, the main reason I didn't use RAII kind of approach was because the compaction can be submit to bottom priority pool in some cases, and it doesn't make sense to throttle it twice.
Nevertheless, after implement TaskLimiterToken and moving the token request procedure during compaction picking, the code look more clear now.